Selecting Parent Blocks: Try clickthrough

jasmussen opened the pull request:
This PR takes a stab at using existing code by @aduth that is implemented on mobile, and expanding it to the desktop breakpoint as well. Specifically, it enables what we in #9628 refer to as "click-through". 

With "click-through", whenever you interact with a block that has innerblocks, a click first select the parent, and then it selects direct descendants. For example you first click the Cover block, then you click the Paragraph block inside. Or you first click the Columns block, then you get to select the Column (singular) block, and then you can select content inside. GIF:

![clickthrough](https://user-images.githubusercontent.com/1204802/57453915-e1fa8380-7267-11e9-960c-2a44230f11dc.gif)

**This branch is a work in progress. There are issues to refine and work out.**

Click-through is inspired by how many desktop applications handle "groups". There, you commonly have to double-click a group to be able to select items inside. This PR is the same, except you click only once to "open" the group. There are a great deal of benefits to this approach:

1. It's very simple. It makes it trivial to select each layer of even complex blocks as they are today. 
2. It's safe, in that it can accommodate almost any complex block you can throw at it, with a few exceptions I'll get into in a bit.
3. It does not require fine motor skills to select a parent block. The hit areas are most literally as big as the block itself. 

The first thing you might think is: **what about situations where you want to edit text quickly?**

That specific question is the primary reason it's taken until now to test this out in practice. My assumption was that this would create unnecessary friction to editing text. However:

- In testing this actual PR, that feels like a complete non-issue. I invite you to test this PR out, even in this buggy early form. 
- The behavior feels completely intuitive in actual practice. Even if you select the parent first, you get visual feedback all along the way, and you'll quickly learn that if you mean to edit text inside the Cover block quickly, you can just double-click — one click to select the Cover block, another to select the text you're pointing at.

**Issues**

1. The current behavior "cycles through" the layers if there are more than two levels of nesting, i.e. as it the case in the Columns block:

![cycle](https://user-images.githubusercontent.com/1204802/57454833-0bb4aa00-726a-11e9-8b39-937b269c4011.gif)

This is the code: `! isSelectedBlockInRoot` — we'll likely need an additional variable that is aware of immediate parent and not just the root ancestor. 

2. There is an issue with blocks being inserted using the slash command — those are still "locked" behind the click-through, even though it is keyboard selected. This will likely be similar to if you set focus inside a child block with the keyboard: however you selected the child block, if it's selected, it should be "ungrouped". 

**Other considerations**

For now this is a draft PR, but I strongly urge you to try it out and _not just look at the GIF_. It feels better in actual practice, than it looks, and with things like #15499 on the horizon, this could be a good baseline improvement to improve the situation with selecting parent and child blocks. 

It also doesn't preclude additional and separate improvements from taking place, such as:

- #14961 which adds dashed outlines to show the presence of parent or child blocks. 
- Refactors of blocks such as columns to maybe be closer to the Media & Text block in terms of how child blocks are surfaced to the user. For example it seems like it should be possible to create a columns block with just 2 levels of nesting, rather than 3.

For now, though, I would really appreciate your thoughts on this PR. Please, more than anything, take it for a spin!

Commit:
Selecting parents: Try clickthrough.

Clickthrough has you select the parent before you can select the child. This is already in place on the mobile breakpoints, this just expands it to desktop as well.

It is a work in progress, right now it is not working as intended: once you have "unlocked" the deepest level, it becomes immediately locked and you have to click through the layers again to unlock it again. The deepest layer should always be unlocked until you deselect all blocks again.

kjellr commented:
🎉 This is an excellent exploration. Like @jasmussen, I originally thought this would seem cumbersome, but in practice it feels really natural. I also agree that it could live alongside a change like #14961: this PR makes it easy to click _down_ the tree of nested blocks, but the other PR would make it easier to click back up the tree as well. I actually tried smashing these PRs together, and in general it seems quite usable to me: 

![borders-too](https://user-images.githubusercontent.com/1202812/57467312-49a4d480-7250-11e9-9162-e74baa0c4f3f.gif)

I 100% agree that clicking 3 times to get into the columns block is annoying, but that block needs some reworking anyway. I think it'd be fine to build in a special case solution for it in the meantime (like maybe the second click bypasses the individual `column` block). 

Aside from the little hiccups Joen noted above there are a couple other aspects of this that need a bit of consideration:

---

1. When a parent block is full-aligned, the hover state is easily missable. (It'll be even more missable if we end up removing the block breadcrumb like we've proposed in other tickets). Since this is the case, it's a little unclear why the child blocks aren't highlighted when you hover over them: 

![full-width](https://user-images.githubusercontent.com/1202812/57465636-05640500-724d-11e9-8e02-f90fb8e7fcce.gif)

This is obviously remedied by just clicking once, so I'm not sure it's a huge deal though. 

---

2. When working in the toolbar of a block that immediately follows a nesting block, toolbar clicks can sometimes select the preceding block, instead of the toolbar item. This appears to happen when the hover state is active for the preceding block: 

![select](https://user-images.githubusercontent.com/1202812/57466068-e619a780-724d-11e9-920f-b375014c83ec.gif)

kjellr commented:
Also, a quick suggestion: in many design apps, you can directly select a child of a group by holding a modifier key when clicking (often the `command` key). It'd be cool to enable that here too for power users.

jasmussen commented:
Thanks so much Kjell, your thoughts are much appreciated, and they are all _solid_. I definitely think the outlines branch and this PR could work well together. It is a very complex UI issue, and multiple approaches are sound. We might even find, if this branch works well, that the outlines branch can become slightly subtler than it is now — thinking about it, it _seems_ perhaps more important to show the outlines when the parent block is selected, given the behavior suggested in this branch.

Your observation on full-wide is solid. I would suspect that the outlines suggested right in the previous paragraph is _one_ way in which this can become more visible. Another might be to actually show a color overlay when hovering to indicate the boundaries. But this is probably best explored separately. 

The 2nd bug you mention, yep I can reproduce it. Great catch, we'll want to fix that for sure. I've asked @gziolo for some help with some of these things, and hopefully we can fix this up together. 

> Also, a quick suggestion: in many design apps, you can directly select a child of a group by holding a modifier key when clicking (often the command key). It'd be cool to enable that here too for power users.

That's a killer tip, I love it. In addition to being able to just double-click, that will be a nice detail!

gziolo committed:
Render overlay on top of inner block only when none of the nested blocks is selected

gziolo commented:
It's better with effe011 but I still need to further investigate how to make this behavior more consistent:

![clickthrough](https://user-images.githubusercontent.com/699132/57631993-d4633780-75a0-11e9-9342-a684d3c08d6f.gif)

I guess it's fine to optimize it with Column block being the one which intercepts the click as it will have to be covered with nested Group blocks regardless. We can remove the overlay behavior from Column block later. That's probably the easiest part.

gziolo commented:
> Also, a quick suggestion: in many design apps, you can directly select a child of a group by holding a modifier key when clicking (often the `command` key). It'd be cool to enable that here too for power users.

I like this proposal a lot. If you look at my screencast you can see that double-click and triple-click trigger the default browser behavior: selecting a word and a group of text.

mapk commented:
I'm loving this PR! Here's some observations:

1. The clickthrough on columns might be a bit confusing to clickthrough 3 times (Columns -> Column -> Content) to get to the content for new users. I think it's something people can grok, but at first clicking to get the column, and not just the content was a bit odd... although I believe it's needed.

2. A BUG: Clicking through to the image in a Columns block pushed me directly to the "Edit" image placeholder. I don't think this should happen, correct?

![clickthrough](https://user-images.githubusercontent.com/617986/57634398-1452fc00-755a-11e9-9ed7-0dd17645c8b9.gif)

The clickthroughs feel right to me. It's so much better than trying to fidget with the cursor to grab just the parent block. Thanks for this exploration!

gziolo commented:
In 5ee69d0 I added background colors to make it easier to debug why a single Column block gets selected from time to time:

![Screen Shot 2019-05-14 at 09 10 37](https://user-images.githubusercontent.com/699132/57677938-521e5600-7628-11e9-9d2b-adb5b699549e.png)

It looks like overlays don't fit properly to the are they should cover.

jasmussen commented:
Thanks everyone for the feedback. I will look a bit today at some fixes to what you've noted:

- Should be able to doubleclick to quickly "unlock" text inside nesting, 
without also selecting text. Maybe possible with some pointer-events magic.
- Will explore an overlay color for the contexts where the hover state is otherwise not obvious, like fullwide. 
- Adjacent toolbars have a lower z index than the overlay — Kjell's reported toolbar issue.

@mapk the Edit Image placeholder thing appears to happen in master too! And as far as I can tell, only in Safari:

![master](https://user-images.githubusercontent.com/1204802/57686995-eabdd180-763a-11e9-951a-03180e666356.gif)

Can you open a new issue for this? It's _possible_ that "double-click" is intended to go to that edit mode, though that feels a bit weird. 

I would also agree that it's a bit annoying that you have to click three times on columns. But I would also suggest that this is a more fundamental issue with the columns block itself, and something that's worth exploring improvements to separately. If at all possible it'd be nice if individual columns didn't have to be selected individually, but that options for resizing and aligning were available directly in the top level columns block. Much like "Media & Text", in other words. 

Given a change was recently merged to make the individual column block selectable again, because it can now be resized, I think we should explore reducing the amount of nesting for this block as a separate effort. What do you think?

jasmussen commented:
In https://github.com/WordPress/gutenberg/pull/15537/commits/84a13f73d09055e1ffc142ca96b7a41b9c8cc712 I fixed the issue where when you try to click the toolbar of a block that's adjacent to a block with the clickthrough overlay, you'd select the overlay. 

But I'm surfacing it here because it feels a bit wrong to be giving the contextual toolbar such a high z-index... but it has to be higher than the click overlay. @aduth can you recall why the click overlay was given a z-index of 120? Do you think 2 could be sufficient?

jasmussen commented:
I spent some time investigating whether it was possible to not _select words_ when doubleclicking, and _select all text_ when triple clicking. As it turns out because this is operating system level behavior, it's not really possible to disable that. You can use various `pointer-events` and `user-select` to prevent selection of text, but once the actual editable text is finally unlocked, you'll still have a triple-click "queued up" which instantly fires to select it.

Barring a hack I cannot think of, this aspect is not fixable without JavaScript. That is — if we truly loathe that quickly double clicking to edit text inside a nested block also selects that text — we'll have to [unset the text selection at the end of the operation](https://stackoverflow.com/questions/10184539/how-to-prevent-the-double-click-select-text-in-javascript). Which we should probably avoid for now, both since it's rare, and because it's a disruptive fix we can explore separately.

kjellr commented:
> Barring a hack I cannot think of, this aspect is not fixable without JavaScript. That is — if we truly loathe that quickly double clicking to edit text inside a nested block also selects that text — we'll have to unset the text selection at the end of the operation. Which we should probably avoid for now, both since it's rare, and because it's a disruptive fix we can explore separately.

Yeah, I think it's okay to leave as is — I generally expect a quick double-click to select text, and I using complicated overrides to change that seems worrisome. Thanks for looking into it.

jasmussen commented:
@kjellr I took a quick stab at your suggestion here:

> When a parent block is full-aligned, the hover state is easily missable. (It'll be even more missable if we end up removing the block breadcrumb like we've proposed in other tickets). Since this is the case, it's a little unclear why the child blocks aren't highlighted when you hover over them:

It looks like this:

<img src="https://cldup.com/i8jYfm_vyD.gif" />

Which is — interesting, but not great. However one thing feels _screamingly_ clear when looking at this to me: this is a use case for your nested borders from #14961! 

In other words: when you hover a block with nesting, a dashed outline is shown about all immediate descendants of the currently locked layer. Right!? To me this feels like a great unification of the two concepts. Simple and contextual.

If you agree, the next question becomes whether we polish this branch up first and then work on those hover borders in a subsequent PR, or whether I take a stab now?

kjellr commented:
It looks like that screenshot isn't showing up for some reason. I'm guessing it's supposed to be the temporary blue overlay in the PR right now, right?

![Screen Shot 2019-05-15 at 8 56 17 AM](https://user-images.githubusercontent.com/1202812/57777218-55c8e000-76ef-11e9-9da1-191a2b4200f6.png)

> In other words: when you hover a block with nesting, a dashed outline is shown about all immediate descendants of the currently locked layer. Right!? To me this feels like a great unification of the two concepts. Simple and contextual.

Is this sort of what you mean? 

![borders-2](https://user-images.githubusercontent.com/1202812/57776897-9ecc6480-76ee-11e9-88dd-e3953ff6c624.gif)

I'm sort of against using these borders on hover: One of the things I like about both the approach in this PR and the one in #14961 is they it keeps the UI nice and quiet until you actually choose to interact with a block. I worry that introducing some block borders on hover will bring us back to the more distracting interface we had before. 

Maybe we could consider adding a simple outer border to full-aligned parent blocks that have no background perhaps, since those are the most problematic:

![borders-2](https://user-images.githubusercontent.com/1202812/57777336-b35d2c80-76ef-11e9-85fc-9836b480c279.gif)

jasmussen commented:
Sorry I commented and didn't verify the image worked. I made it a link now, and if you click it it works.

It's hard to convey exactly why it feels right to me, and yes that GIF you recorded is sort of in the right direction. Though it would need some finetuning — which suggests to answer the question at hand, that it should probably be expored separately. Maybe it will naturally come up if this PR gets merged first and we have to rebase yours. 

I think there's potential there, though!

jasmussen commented:
I would love to ship an MVP of this PR, this week. What are the next steps?

Honestly I feel like if we remove the debug overlays and ship as is, and then ship an MVP of Kjell's nested block outlines shortly after, we will be in a VASTLY better place than we are now. Then after that perhaps we can revisit in separate small PRs ideas that have been discussed:

- On-select padding
- "opt-out"
- tweaked hover styles (perhaps revisit the block label) to better indicate hover?

Let me know what you think are the next steps! cc @youknowriad @gziolo @kjellr

gziolo commented:
I took it for another spin and it works great overall. There is some issue with the left side padding in the overlay. There is some area where you are able to select a nested block directly. I think it should be fixed before we merge this PR. I would vote to land this proposal 👍 

![clickthrough](https://user-images.githubusercontent.com/699132/58020411-4d661000-7b08-11e9-8e83-0500b827d6e3.gif)

gziolo commented:
There is a related issue with block selection caused by all block wrappers. It's hard to reproduce but I think I recorded it at the end of this screencast:
![block-wrapper](https://user-images.githubusercontent.com/699132/58020603-e1d07280-7b08-11e9-94b7-d60709e6f9b9.gif)

gziolo commented:
> There is a related issue with block selection caused by all block wrappers. It's hard to reproduce but I think I recorded it at the end of this screencast:
> ![block-wrapper](https://user-images.githubusercontent.com/699132/58020603-e1d07280-7b08-11e9-94b7-d60709e6f9b9.gif)

This exists in `master` and could be improved separately:
![Screen Shot 2019-05-20 at 14 53 56](https://user-images.githubusercontent.com/699132/58023103-40005400-7b0f-11e9-8263-28d29b5e5b88.png)

This better illustrates the issue. This flickering of the block's title is annoying as well. It might be related to the animation it gets. I'm still not sure why it shows up and hides from time to time even though the left side border is there all the time.

Commit:
Fix overlay, fix breadcrumb, polish.

jasmussen commented:
Alright friends, I've done the following:

- Rebased/squashed to keep this up to snuff with master.
- Fixed an issue where a nested breadcrumb would be covered by both overlay and toolbars.
- Fixed the issue reported where the overlay on the Column block pulled out to either side and was selectable even if it wasn't yet "unlocked". Fixed this in the columns block styles itself as it's limited to the margin voodoo that happens there. 
- Removed debug stuff, and reduced opacity of the overlay. 

GIF:

![status](https://user-images.githubusercontent.com/1204802/58079162-abe6c900-7bb0-11e9-99ac-24e8318307bd.gif)

I would love your sanity checks on the following:

1. Should we ditch the overlay color? It doesn't feel like it's adding much value, especially compared to Kjell's block outlines. Should we maybe merge with the overlay present, but remove it when we rebase Kjell's PR?
2. Can I get a sanity check on the new z-index for nested breadcrumbs?

Would love to ship this after. Thanks all.

gziolo commented:
Code wise this PR looks good. We probably will have to update 2 failing unit tests:

https://travis-ci.com/WordPress/gutenberg/jobs/201739741#L1086
https://travis-ci.com/WordPress/gutenberg/jobs/201739745#L1088

```
FAIL packages/e2e-tests/specs/reusable-blocks.test.js (54.43s)
  ● Reusable Blocks › can be created from multiselection
    waiting for XPath "//button[text()="Edit"]" failed: timeout 30000ms exceeded
      238 | 
      239 | 		// Wait for saving to finish
    > 240 | 		await page.waitForXPath( '//button[text()="Edit"]' );
          | 		           ^
      241 | 
      242 | 		// Check that we have a reusable block on the page
      243 | 		const block = await page.$( '.block-editor-block-list__block[data-type="core/block"]' );
      at new WaitTask (../../node_modules/puppeteer/lib/FrameManager.js:840:28)
      at Frame._waitForSelectorOrXPath (../../node_modules/puppeteer/lib/FrameManager.js:731:12)
      at Frame.waitForXPath (../../node_modules/puppeteer/lib/FrameManager.js:699:17)
      at Page.waitForXPath (../../node_modules/puppeteer/lib/Page.js:1017:29)
      at waitForXPath (specs/reusable-blocks.test.js:240:14)
      at tryCatch (../../node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (../../node_modules/regenerator-runtime/runtime.js:271:22)
      at Generator.prototype.(anonymous function) [as next] (../../node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (specs/reusable-blocks.test.js:33:103)
      at _next (specs/reusable-blocks.test.js:35:194)
  ● Reusable Blocks › multi-selection reusable block can be converted back to regular blocks
    TypeError: Cannot read property 'click' of undefined
      at _callee$ (../e2e-test-utils/build/@wordpress/e2e-test-utils/src/insert-block.js:13:8)
      at tryCatch (../../node_modules/regenerator-runtime/runtime.js:45:40)
      at Generator.invoke [as _invoke] (../../node_modules/regenerator-runtime/runtime.js:271:22)
      at Generator.prototype.(anonymous function) [as next] (../../node_modules/regenerator-runtime/runtime.js:97:21)
      at asyncGeneratorStep (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (../../node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
```

They failed on 2 jobs so I predict that this will require some rework. I can take care of it tomorrow so you all can focus on manual testing and design aspects :)

kjellr commented:
I'll give this a test later today, after I clean up #14961. In the meantime: 

> Should we ditch the overlay color? It doesn't feel like it's adding much value, especially compared to Kjell's block outlines. Should we maybe merge with the overlay present, but remove it when we rebase Kjell's PR?

Yep, I say we remove the overlay for now. We can always add it again in the future, but we'll have a better sense of its usefulness + design objective once we have this PR in alongside the borders + padding.

jasmussen committed:
Remove click-overlay.

jasmussen commented:
Removed the overlay. Thanks, Gzregorz, for help with the tests!

kjellr commented:
Gave this a test now, and I'm actually noticing that the clickthrough doesn't _always_ work for me. If I click through to a child, and then deselect completely, I can sometimes click directly back into the same child:

![Untitled 2019-05-21 10_07_14](https://user-images.githubusercontent.com/1202812/58103089-60341f80-7bb0-11e9-8a5a-e46e0e5a61d3.gif)

I haven't found any distinct steps to reproduce, but it sometimes seems to relate to speed: If I click off a child block and then click back into it quickly, it allows me to select it. If I wait a couple seconds and try to click back in, I have to click all the way back down from the top level. Maybe this is intended?

jasmussen commented:
Amazing catch Kjell, thank you. I will take a look, but also CC: @aduth because he wrote the initial code for this, in case he has context.

jasmussen commented:
So, I took a deep dive on the overlay issue. There's definitely a bug there, and I think it may be beyond my skill to fix. 

The issue is not present in either Chrome or Firefox, but is trivially reproducible in Safari: simply click outside a nested block to deselect it, and the overlay is not restored. 

I added the following code to `packages/block-editor/src/components/inner-blocks/index.js`:

```
console.log( 'Should have overlay: ' + ( ! isBlockSelected( clientId ) && ! hasSelectedInnerBlock( clientId, true ) ) );
```

☝️this is the code that decides whether `hasOverlay` is true or false. Here's how it looks in Chrome:

![chrome](https://user-images.githubusercontent.com/1204802/58157865-88d41c00-7c79-11e9-9968-f1a2381057b0.gif)

As you can see, it seems _mostly_ consistent. Overlays are applied and then the loop stops. Once the final layer is unlocked, the loop rests at "false", which is correct because the overlay should not be present. And it reverts to true when the block is deselected.

Now watch what happens in Safari:

![saf](https://user-images.githubusercontent.com/1204802/58158006-d781b600-7c79-11e9-9748-6e56cd1021d1.gif)

Even doing nothing, it logs multiple times. It seems to rest at 6, but it's weirdly slow compared to Chrome where those overlays are more or less instantiated instantly. Something similar happens when clicking — the overlays are slowly instantiated, but it continues to count to 6 or so. When unlocking the final layer, though, something goes awry and instead of resting on "false" because the overlay is now unlocked, it flips back and forth. 

@gziolo and @aduth, any insights here would be deeply appreciated.

jasmussen commented:
Noting that the code in https://github.com/WordPress/gutenberg/pull/14961/files#r284278164 fixes the issue described right above. I'm aware that the code suggested should not land, but sharing as a data point.

kjellr commented:
> Noting that the code in https://github.com/WordPress/gutenberg/pull/14961/files#r284278164 fixes the issue described right above. I'm aware that the code suggested should not land, but sharing as a data point.

Ah, interesting. Well the good news is that needs to be fixed for _both_ PRs then. 😄

aduth commented:
> But I'm surfacing it here because it feels a bit wrong to be giving the contextual toolbar such a high z-index... but it has to be higher than the click overlay. @aduth can you recall why the click overlay was given a z-index of 120? Do you think 2 could be sufficient?

I don't immediately recall. The preceding comment may be the necessary context? `120` may have been chosen because `.components-drop-zone__content` has a `z-index` of `110`.

https://github.com/WordPress/gutenberg/blob/84a13f73d09055e1ffc142ca96b7a41b9c8cc712/assets/stylesheets/_z-index.scss#L60-L61

(Originally introduced in #6773)

aduth commented:
> So, I took a deep dive on the overlay issue. There's definitely a bug there, and I think it may be beyond my skill to fix.

Still debugging on my end as well.

A few notes:

- To reproduce in Safari, the important thing seems to be to deselect from an activated inner block.
   1. Insert Columns block
   2. Insert a block (e.g. Image) in the first Column
   3. Save and reload for good measure
   4. Click to select the Columns
   5. Click again to select Column
   6. Click in the gutter aside the block list
   7. Click on the Column
      - It should be expected that Columns is selected, but instead Column is
- Your `console.log` snippet is going to log for any state change anywhere, and also for any InnerBlocks (including other Column block), so it might be a bit difficult to decipher where and for what reason changes are actually happening.
- That this occurs in Safari specifically is confusing.
- Just in case, I verified there aren't Safari fixes in [latest releases of React](https://github.com/facebook/react/blob/master/CHANGELOG.md).
- The state itself appears correct at the point where the overlay is missing, such that the selectors produce accurate values which would indicate that it _shoulnd't_ be missing. Maybe there's something specific here with React reconciling and Safari.
- Could also be worth investigating the specific [implementation of `BlockSelectionClearer](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-selection-clearer/index.js) (i.e. `tabIndex`). It's not the first time events for focusable elements have been unpredictable in Safari ([example](https://github.com/WordPress/gutenberg/blob/f1bbc004734ea157dd72ed5bf1f808c10efc0e84/packages/block-editor/src/components/block-list/insertion-point.js#L61-L66))

aduth commented:
While it might be considered more an incidental fix to the underlying issue, as noticed in #15499, for the sake of end-to-end tests and general usability, it might be worth considering to disable async rendering when the selected block occurs in the same block hierarchy. Currently the logic only considers whether the block is selected.

https://github.com/WordPress/gutenberg/blob/f1bbc004734ea157dd72ed5bf1f808c10efc0e84/packages/block-editor/src/components/block-list/index.js#L205-L213

Not entirely sure it would fix the issue, but worth exploring.

Alternatively, we could do some browser sniffing to disable async rendering for Safari altogether if we suspect it's not stable enough due to some browser-specific circumstances.

cc @youknowriad

youknowriad commented:
@aduth

I agree we can try to tweak the condition to account for nesting but we need to ensure two things:

 - Adding the condition (new selectors) shouldn't impact performance
 - Right now, most of the blocks are not in a nested context but what happens when everything is a block including the layout, we'll end up with almost all blocks in a nested context, which means, almost all the blocks will be in sync mode everytime and we'd lose the benefit of the async mode.
 - I don't think we should have differences between e2e tests and prod environments regarding the async mode to avoid hidden bugs.

So yeah, let's explore tweaking that condition but we should check the performance impact on long posts.

gziolo committed:
Fix the selection of inner blocks for reusable blocks template

gziolo commented:
b66070a - should fix the failing e2e tests. It isn't perfect but does the job perfectly fine. We should seek for refactoring for reusable blocks to use nested block editor instead to fix all the issues I could see when testing. It's out of the scope of this PR and definitely not a blocker:

![reusable-blocks](https://user-images.githubusercontent.com/699132/58421424-2f605880-8090-11e9-9166-1a3e846a484d.gif)
![Screen Shot 2019-05-27 at 14 03 11](https://user-images.githubusercontent.com/699132/58421425-2f605880-8090-11e9-8f36-cde87a3d8db0.png)

gziolo commented on packages/block-editor/src/components/inner-blocks/index.js:
```diff
@@ -139,12 +139,13 @@ InnerBlocks = compose( [
+			block,
 			blockListSettings: getBlockListSettings( clientId ),
-			hasOverlay: ! isBlockSelected( clientId ) && ! hasSelectedInnerBlock( clientId, true ),
+			hasOverlay: block.name !== 'core/template' && ! isBlockSelected( clientId ) && ! hasSelectedInnerBlock( clientId, true ),
```
@youknowriad - any thoughts on this one? It's a really nasty hack to avoid having the overlay in the context of nesting...

youknowriad replied:
I think it's fine, ideally, this `core/template` block should be removed. Rebuilding the reusable blocks UI as a nested editor would allow us to remove it. See #14367

That PR is blocked by the fact that it's hard to use nested Slot/Fills. One tradeoff we can make to unblock it could be to decide that we don't show the inspector settings for the reusable blocks content and redirect the user to the edit page of the reusable blocks for a full editing session.

gziolo committed:
Disable async mode for parent blocks when nested block selected

gziolo commented:
I'm experimenting with enforcing sync mode in the block which contains selected nested block. 73fa2a0 improves it only slightly. I discovered during testing that it gets applied only to parent blocks but it should get applied to all blocks which share the same hierarchy root client id.

![sync-mode](https://user-images.githubusercontent.com/699132/58424345-83bb0680-8097-11e9-9c3f-4bc0055a400a.gif)

I will continue tomorrow.

gziolo commented on packages/block-editor/src/components/block-list/index.js:
```diff
@@ -204,14 +205,15 @@ class BlockList extends Component {
 							value={ ! isBlockInSelection }
 						>
 							<BlockListBlock
+								className={ isBlockInSelection ? 'is-sync' : 'is-async' }
```
It's temporary to make debugging easier :)

gziolo commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -209,6 +209,10 @@
  */
 
 .block-editor-block-list__layout .block-editor-block-list__block {
+	&.is-sync {
```
Added only for debugging purposes.

gziolo commented:
I'm not sure if this is only this branch or an issue in general but it's nearly impossible to drag or move the block in 2nd and the following columns: 
![issue-with-movers](https://user-images.githubusercontent.com/699132/58454232-19997480-811e-11e9-8fde-701bb958da94.gif)

gziolo committed:
Refactor BlockListBlock to use AsyncModeProvider

gziolo commented:
I moved `AsyncModeProvider` one level down with https://github.com/WordPress/gutenberg/pull/15537/commits/cfaf400afb0ca666a1eb2ff8ea5944c72152f7e0. This makes it much easier to decide which blocks should have sync mode forced. In fact, we don't have to use any additional props to make it work as expected (unless I missed something) and even can remove 3 props from the `BlockList` component.

youknowriad commented:
@gziolo I fear this defeats the whole purpose of the async mode. It's there to avoid running all the `BlockListBlock` selectors on each key strike. You can run the performance tests on the branch to confirm.

jasmussen commented:
Thanks, Gzregorz, for all your help. I tested the most recent changes you made, and clickthrough wise it feels like an improvement. Safari is still not perfect — if you click fast enough in some situations you can still avoid the overlay. But it does feel like it's good enough to ship, user experience wise. 

The bad news is that it still is a little flimsy, and if it causes the isseus Riad mentions, then it may need a little more time in the oven. I would really appreciate everyone help Gzregorz with this — selecting nested blocks is critical path stuff for phase 2, and this PR along with Kjell's outlines changes which relies upon the same async fix, is currently our best bet. So whatever technical changes we need to fix, would appreciate all the eyes on it. 

I noticed an additional weirdness -- for every time you click an overlay, there's additional markup being created:

![new markup](https://user-images.githubusercontent.com/1204802/58460587-4786b500-812e-11e9-9682-9fe0fba815e4.gif)

Not entirely sure what happens here.

---

The reusable blocks UI is also not accessible in this branch. You can click the edit button, but not the save button. I can look into this, but per comments in https://github.com/WordPress/gutenberg/pull/15537#issuecomment-496205232 I'm unsure how much to do?

Let me know.

jasmussen committed:
Make the reusable blocks save button clickable. i

At least this means you can edit and save reusable blocks.

jasmussen committed:
Fix so reusable blocks with nesting are editable.

jasmussen commented:
I pushed a few fixes so reusable blocks function as intended.

jasmussen commented:
> I fear this defeats the whole purpose of the async mode. It's there to avoid running all the BlockListBlock selectors on each key strike. You can run the performance tests on the branch to confirm.

@youknowriad can you elaborate a little bit on this? Do the performance issues occur when blocks with nesting only have a few blocks, or is it only the case when there are many, many blocks nested?

I would repeat that a method to select nested blocks is very important to solve, and the issue that Gzregorz has uncovered with the async mode is not only affecting this branch, but Kjell's too. We need a path forward. It works reasonably well in Chrome and Firefox, but is completely broken in Safari. I don't mean we should sacrifice performance, but I do believe that this PR uncovers issues present in master even today. See some of the debugging done in https://github.com/WordPress/gutenberg/pull/15537#issuecomment-494701888 — I wonder how it affects other aspects.

gziolo commented:
> I fear this defeats the whole purpose of the async mode. It's there to avoid running all the `BlockListBlock` selectors on each key strike. You can run the performance tests on the branch to confirm.

Okay, I think I understand your concerns. Would it make it better it `AsyncModeProvider` would be moved to the new parent component?

jasmussen committed:
Fix movers.

The z-index was too low. It had to be higher to accommodate the other z-index changes.

jasmussen commented:
I pushed a fixed to the movers to make them functional again. Outside of any async fixes necessary, which I will defer to all of you experts (thank you 💖) — this branch should be good to go.

youknowriad commented:
> @youknowriad can you elaborate a little bit on this? Do the performance issues occur when blocks with nesting only have a few blocks, or is it only the case when there are many, many blocks nested?

@jasmussen This is about the number of blocks regardless of whether we have nested blocks or not.

> I would repeat that a method to select nested blocks is very important to solve, and the issue that Gzregorz has uncovered with the async mode is not only affecting this branch, but Kjell's too

I know but just to clarify a little bit, that's an issue, it's how it's meant to work. Changes to unselected blocks are executed async in order to avoid holding the typing and the feeling of reactiveness. I do think that's way more important.

That said, I think there are probably ways to solve it in these branches by saying something like: "Direct parents can also rerender synchronously" or "direct children can rerender synchronously". We also need to think of this in regards of the "group block" or Full site editing where everything will be nested. So we can just say if your root block is selected or there's a selected block within the same root block, rerender synchronously because in that case, we'll be just rerendering everything since all blocks can will most likely have the same parent.

> Okay, I think I understand your concerns. Would it make it better it AsyncModeProvider would be moved to the new parent component?

@gziolo I'm not certain what parent component you're referring too but the most important thing is that if a post contain a list of 100 blocks at the root level, the AsyncModeProvider should be rendered only once.

gziolo commented:
> That said, I think there are probably ways to solve it in these branches by saying something like: "Direct parents can also rerender synchronously" or "direct children can rerender synchronously". We also need to think of this in regards of the "group block" or Full site editing where everything will be nested. So we can just say if your root block is selected or there's a selected block within the same root block, rerender synchronously because in that case, we'll be just rerendering everything since all blocks can will most likely have the same parent.

This is exactly what I'm trying to achieve in this branch. At the moment sync updates are applied to:
- selected block
- multiselected blocks
- all parents of the selected block

> @gziolo I'm not certain what parent component you're referring too but the most important thing is that if a post contain a list of 100 blocks at the root level, the AsyncModeProvider should be rendered only once.

It is rendered for every `BlockListBlock` in `master` 🤷‍♂: 

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-list/index.js#L204-L223

youknowriad commented:
> It is rendered for every BlockListBlock in master 🤷‍♂:

Right sorry, I meant the selectors used to render it should only be run once. That said, the performance tests are the source of truth here :)

jwold commented:
From https://github.com/WordPress/gutenberg/issues/12515:

> What I do like about this issue is the concept of making the block labels clickable as in a breadcrumb system.

I really like that idea!

gziolo committed:
Bring the behavior closer to what we have as of today

gziolo commented:
## Performance benchmarks

This branch:

```Average time to load: 5503ms
Average time to DOM content load: 4805ms
Average time to type character: 98.98ms
Slowest time to type character: 244ms
Fastest time to type character: 72ms```

```
Average time to load: 5831ms
Average time to DOM content load: 5278ms
Average time to type character: 110.195ms
Slowest time to type character: 266ms
Fastest time to type character: 77ms```

```
Average time to load: 5801ms
Average time to DOM content load: 5142ms
Average time to type character: 99.555ms
Slowest time to type character: 218ms
Fastest time to type character: 72ms```

gziolo reviewed (approved):
It looks good code-wise. I added most of the JavaScript changes so I would appreciate confirmation from @youknowriad given his expertise around the performance implications related to changes applied. Benchmark results look good but I might miss something important.

youknowriad commented on packages/block-editor/src/components/block-list/block-async-mode-provider.js:
```diff
@@ -0,0 +1,23 @@
+	useSelect,
+} from '@wordpress/data';
+
+const BlockAsyncModeProvider = ( { children, clientId, isBlockInSelection } ) => {
```
Nit: Conceptually speaking and if there's no impact on performance it would be better to avoid the `isBlockInSelection` as a prop and compute it here :).

gziolo replied:
https://github.com/WordPress/gutenberg/pull/15537/commits/b0649e1f9b4d846a191993918967e20619a6d907 - I had it before but I assumed it could impact performance. I can try with the previous commit.

gziolo replied:
I don't thin that benchmark agrees with this idea: 
```
Average time to load: 5724ms
Average time to DOM content load: 5219ms
Average time to type character: 187.28ms 😅 
Slowest time to type character: 319ms
Fastest time to type character: 158ms
```

youknowriad replied:
:)

jasmussen commented:
Hooraay! I see a green button and approvals. I will wait for final sanity checks by Riad, and advice on when it's best to merge. But I'm excited to see this land. It solves numerous ux issues around block nesting, and although it takes a little getting used to first time you use it, the PR that @kjellr has pending shows deftly that it is possible to improve the UX further, with visual feedback. Thanks everyone.

youknowriad reviewed (approved):
Run the performance tests as well and I have similar results as well.

youknowriad commented:
Thanks everyone for your hard work on this PR 👏

jasmussen commented:
Hooray! Thank you all for the help 

Now that it's in, I will work with Kjell to refine the experience.

mapk commented:
This PR rocks! Thanks for continued push to merge. You all did fantastic here! 🎉

niklasp commented:
nice. i want to use it now.

ktmn commented:
@jasmussen Thanks, I hate it :)

I can see this feature being useful for the average user, but as a "power user" it slows me down and confuses me immensely. In some instances I need to click up to 5 times just to edit some text. I guess everyone else only tested it with core blocks and 1-2 levels of nesting, in which case it's fine.

Additionally now it's hard to select an element with the Dev Tools since it won't let you select it unless your editor focus is in the right place.

**Definately needs a toggle in my opinion.**

I would prefer a feature such as holding down `Alt` and scrolling up to select parent block and scrolling down to cycle between children blocks.

jasmussen commented:
Thanks for your feedback. There are a number of refinements to this underway, including some outlines to show parent and child elements and their relationship when a block is selected, and it's also been suggested to implement a keyboard shortcut, so when you hold for example ⌘ and click, you click the nested block under your mouse directly. 

A per-block opt-in or opt-out has also been discussed. This would be set by the block author, to opt in or out of the click overlays. So for example if the author of the Media & Text block might opt out of the overlay because it's not necessary for the operation of the block. However for other blocks such as the Columns block, it's virtually impossible to select the individual columns without this interface, so it's important to have a baseline tool for making that possible even for casual users that don't know about keyboard shortcuts or toolbar buttons. 

I'm personally excited about the keyboard shortcut. It feels like a good balance between a userfriendly interface for the casual user, and a tool for powerusers to know if they want to click right through it all. @gziolo do you have any thoughts on how easy/hard such a keyboard shortcut would be to build? Essentially the clickthrough layer needs to be aware of whether the button is pressed down, and then simply behave as if it's not there (`pointer-events: none;`).

gziolo commented:
> I'm personally excited about the keyboard shortcut. It feels like a good balance between a userfriendly interface for the casual user, and a tool for powerusers to know if they want to click right through it all. @gziolo do you have any thoughts on how easy/hard such a keyboard shortcut would be to build? Essentially the clickthrough layer needs to be aware of whether the button is pressed down, and then simply behave as if it's not there (`pointer-events: none;`).

We use a similar approach for multi-selecting blocks when you hold down `Shift` key and click on blocks. I'm sure we can replicate a similar behavior for the clickthrough layer. I think this is the existing logic:

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/writing-flow/index.js#L301-L319

gziolo commented:
I did a quick check with the following code:
```diff
diff --git a/packages/block-editor/src/components/inner-blocks/index.js b/packages/block-editor/src/components/inner-blocks/index.js
index 75a6731f4..bc59bd92f 100644
--- a/packages/block-editor/src/components/inner-blocks/index.js
+++ b/packages/block-editor/src/components/inner-blocks/index.js
@@ -29,9 +29,12 @@ class InnerBlocks extends Component {
 	constructor() {
 		super( ...arguments );
 		this.state = {
+			clickthroughEnabled: true,
 			templateInProcess: !! this.props.template,
 		};
 		this.updateNestedSettings();
+		this.onKeyDown = this.onKeyDown.bind( this );
+		this.onKeyUp = this.onKeyUp.bind( this );
 	}
 
 	getTemplateLock() {
@@ -102,20 +105,42 @@ class InnerBlocks extends Component {
 		}
 	}
 
+	onKeyDown( event ) {
+		if ( event.metaKey ) {
+			console.log( event.metaKey );
+			this.setState( {
+				clickthroughEnabled: false,
+			} );
+		}
+	}
+
+	onKeyUp( event ) {
+		if ( event.metaKey ) {
+			console.log( event.metaKey );
+			this.setState( {
+				clickthroughEnabled: true,
+			} );
+		}
+	}
+
 	render() {
 		const {
 			clientId,
 			hasOverlay,
 			renderAppender,
 		} = this.props;
-		const { templateInProcess } = this.state;
+		const { clickthroughEnabled, templateInProcess } = this.state;
 
 		const classes = classnames( 'editor-inner-blocks block-editor-inner-blocks', {
-			'has-overlay': hasOverlay,
+			'has-overlay': hasOverlay && clickthroughEnabled,
 		} );
 
 		return (
-			<div className={ classes }>
+			<div
+				className={ classes }
+				onKeyDown={ this.onKeyDown }
+				onKeyUp={ this.onKeyUp }
+			>
 				{ ! templateInProcess && (
 					<BlockList
 						rootClientId={ clientId }
```

I don't think it works as expected, this might need to have to be controlled from `WritingFlow` as well as other keyboard events. It definitely won't work with `Shift` as it conflicts with multiselection, but it might be fine to use ⌘ and click as suggested above :+1:

jasmussen commented:
Awesome, thanks so much for looking.

We definitely want to use ⌘ on a Mac, as that's the standard keyboard shortcut for this behavior. Figma, Sketch and Illustator are examples of this behavior, where if you have a group, a single click selects it, but a ⌘ click selects the item inside the group directly.

I can't recall what the corresponding Windows shortcut key is, it's probably Alt, but perhaps @kjellr can recall?

kjellr commented:
> I can't recall what the corresponding Windows shortcut key is, it's probably Alt, but perhaps @kjellr can recall?

Alt sounds right to me, but I don't have a windows machine around to double check at the moment.
